connect/runtime/src/main/java/org/apache/kafka/connect/runtime/PredicatedTransformation.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.runtime;
+
+import java.util.Map;
+
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.connect.connector.ConnectRecord;
+import org.apache.kafka.connect.transforms.Transformation;
+import org.apache.kafka.connect.transforms.predicates.Predicate;
+
+/**
+ * Decorator for a {@link Transformation} which applies the delegate only when a
+ * {@link Predicate} is true (or false, according to {@code negate}).
+ * @param <R>
+ */
+class PredicatedTransformation<R extends ConnectRecord<R>> implements Transformation<R> {
+
+    /*test*/ final Predicate<R> predicate;
+    /*test*/ final Transformation<R> delegate;
+    /*test*/ final boolean negate;
+
+    PredicatedTransformation(Predicate<R> predicate, boolean negate, Transformation<R> delegate) {
+        this.predicate = predicate;
+        this.negate = negate;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+
+    }
+
+    @Override
+    public R apply(R record) {
+        if (negate ^ predicate.test(record)) {
+            return delegate.apply(record);
+        }
+        return record;
+    }
+
+    @Override
+    public ConfigDef config() {
+        return null;
```

Comment:
This is related to the issue discussed about `configure()`.

We could return an empty `ConfigDef` here, but that would be a lie which could ultimately lead some other error if someone tried to use it with `configure()`.

We can't invent a `ConfigDef` schema for this because the `PredicatedTransformer` would need to know about the `Transformer` is was going to be wrapping, but it can't know that before it's been configured with at least the `Transformer`'s `ConfigDef` and it can't be configured before `config()` has been called. So we have a chicken and egg problem. Something (`ConnectorConfig`) must have some _a priori_ knowledge of either `PredicatedTransformer`'s `ConfigDef`, or know how to configure it without needing to call `config()` at all.

Since `PredicatedTransformer` is a purely internal class which will never be directly exposed to Connect users, we're not obliged to stick to the contract of `config()` and `configure()`. i.e. So  both `PredicatedTransformer.config` and `PredicatedTransformer.configure` can throw when called, since we know no one else can call them and we know `ConnectorConfig` never will.

Reply:
Hmmm... I think there might be some awkwardness here with trying to make `PredicatedTransformer` implement the `Transformation` interface. Could we replace every `Transformation` in the `TransformationChain`'s [transformation list](https://github.com/apache/kafka/blob/de6468ae5915298279e229dc64721e01e7d14fab/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java#L33) with a `PredicatedTransformer` and, if there are no predicates configured for a transform by the user, make the default behavior for the `PredicatedTransformer` class to blindly apply its transformation?

This would solve a few problems:

- No risk of users trying to actually use a `PredicatedTransformer` in a connector config, which they may try to do if we don't add logic to prevent it from being picked up during plugin path scanning on startup and logged as an SMT plugin
- No need to implement methods that aren't used
- One code path instead of two for application of transformations
- More flexibility in instantiation and, possibly, the ability to encapsulate some of the `ConfigDef` generation logic in a separate class from `ConnectorConfig` (haven't looked into the specifics of this yet so may not actually be feasible or that elegant)